lib/tasks/data/truncate_data.rb

```diff
@@ -0,0 +1,95 @@
+# frozen_string_literal: true
+
+class TruncateData
+  def initialize(months_to_keep: nil)
+    @date = (months_to_keep || 3).months.ago
+  end
+
+  def call
+    truncate_inventory
+    truncate_adjustments
+    truncate_order_associations
+    truncate_order_cycle_data
+
+    sql_delete_from "spree_orders #{where_oc_id_in_ocs_to_delete}"
+
+    truncate_subscriptions
+
+    sql_delete_from "order_cycles #{where_ocs_to_delete}"
+
+    Spree::TokenizedPermission.where("created_at < '#{date}'").delete_all
+
+    remove_transient_data
+  end
+
+  private
+
+  attr_reader :date
+
+  def truncate_order_associations
+    sql_delete_from "spree_line_items #{where_order_id_in_orders_to_delete}"
+    sql_delete_from "spree_payments #{where_order_id_in_orders_to_delete}"
+    sql_delete_from "spree_shipments #{where_order_id_in_orders_to_delete}"
+  end
+
+  def remove_transient_data
+    Spree::ReturnAuthorization.delete_all
+    Spree::StateChange.delete_all
+    Spree::LogEntry.delete_all
+    sql_delete_from "sessions"
```

Comment:
I am not sure we can just wipe this out for production.

Spree::ReturnAuthorization
this is not something we can do it prod, although the last time I checked with had under 10 records across all DBs.
There must be a reason for this to be here... I wonder if this is necessary for the staging process, otherwise we could remove this one from the task.

The othres, sessions, StateChange and LogEntry we should probably add a date filter to the delete statement.
,

Reply:
Let me go through what think of each and see if we find a sweet spot:

* No idea what ReturnAuthorizations is. Something related to returned shipments? we need to investigate.
* StateChange: since this is useful for troubleshooting a month data seems enough.
* LogEntry: the same
* Sessions: this affects users that actively purchasing so I think 2 weeks data is more than enough. Others can afford to log in again.

Reply:
I'm going to go with this.

Reply:
there are only 5 spree_return_authorizations in aus live but we cant just delete them, there is one in 2020.
My code is a lazy shortcut, the ret auth are linked to orders.
So all you need to do is:
      sql_delete_from "spree_turn_authorizations" #{where_order_id_in_orders_to_delete}"

I havent tested but I think this will do it :+1:

Reply:
I just check what these authorizations are. Yep, your suggestion makes sense.